File: crates/traderview-db/src/rmd_calculator.rs
//! Required Minimum Distribution (RMD) calculator.
//!
//! Per the SECURE 2.0 Act (2022) and IRS Publication 590-B:
//!
//!   - RMDs begin at age 73 for those born 1951-1959 (and at age 75
//!     for those born 1960+).
//!   - RMD_year_x = prior_year_end_balance / IRS_Uniform_Lifetime_Factor_age_x
//!   - Missed RMD penalty (SECURE 2.0): 25% of shortfall (reduced from
//!     50% pre-2023), or 10% if corrected within 2 years.
//!
//! This module embeds the IRS Uniform Lifetime Table (effective 2022,
//! current as of Publication 590-B 2024 update) for ages 72-120. Also
//! supports projection over N years assuming the user's balance grows
//! at `expected_return_pct` net of the RMD withdrawal each year.
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct RmdInput {
    pub birth_year: i32,
    pub current_age: u32,
    pub balance_usd: f64,
    #[serde(default = "default_return")]
    pub expected_annual_return_pct: f64,
    /// Years to project forward.
    #[serde(default = "default_proj_years")]
    pub project_years: u32,
}

fn default_return() -> f64 { 6.0 }
fn default_proj_years() -> u32 { 20 }

#[derive(Debug, Clone, Serialize)]
pub struct YearRmd {
    pub age: u32,
    pub start_balance_usd: f64,
    pub rmd_factor: f64,
    pub rmd_amount_usd: f64,
    pub end_balance_after_rmd_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct RmdReport {
    pub rmd_start_age: u32,
    pub current_factor: Option<f64>,
    pub current_rmd_usd: Option<f64>,
    pub years_until_rmd: i32,
    pub projection: Vec<YearRmd>,
    pub total_rmds_through_projection_usd: f64,
}

// ─── IRS Uniform Lifetime Table (2022) ────────────────────────────────────

/// Divisor for age `a`. Outside the table range, returns None.
/// Table source: IRS Publication 590-B, Appendix B, Table III.
pub fn uniform_lifetime_factor(age: u32) -> Option<f64> {
    let f = match age {
        72 => 27.4,
        73 => 26.5,
        74 => 25.5,
        75 => 24.6,
        76 => 23.7,
        77 => 22.9,
        78 => 22.0,
        79 => 21.1,
        80 => 20.2,
        81 => 19.4,
        82 => 18.5,
        83 => 17.7,
        84 => 16.8,
        85 => 16.0,
        86 => 15.2,
        87 => 14.4,
        88 => 13.7,
        89 => 12.9,
        90 => 12.2,
        91 => 11.5,
        92 => 10.8,
        93 => 10.1,
        94 => 9.5,
        95 => 8.9,
        96 => 8.4,
        97 => 7.8,
        98 => 7.3,
        99 => 6.8,
        100 => 6.4,
        101 => 6.0,
        102 => 5.6,
        103 => 5.2,
        104 => 4.9,
        105 => 4.6,
        106 => 4.3,
        107 => 4.1,
        108 => 3.9,
        109 => 3.7,
        110 => 3.5,
        111 => 3.4,
        112 => 3.3,
        113 => 3.1,
        114 => 3.0,
        115 => 2.9,
        116 => 2.8,
        117 => 2.7,
        118 => 2.5,
        119 => 2.3,
        120 => 2.0,
        _ => return None,
    };
    Some(f)
}

pub fn rmd_start_age_for_birth_year(birth_year: i32) -> u32 {
    if birth_year >= 1960 { 75 }
    else if birth_year >= 1951 { 73 }
    else { 72 }
}

pub fn compute(input: &RmdInput) -> RmdReport {
    let start_age = rmd_start_age_for_birth_year(input.birth_year);
    let years_until = start_age as i32 - input.current_age as i32;

    let current_factor = if input.current_age >= start_age {
        uniform_lifetime_factor(input.current_age)
    } else { None };
    let current_rmd = current_factor.map(|f| input.balance_usd / f);

    let r = input.expected_annual_return_pct / 100.0;
    let mut balance = input.balance_usd;
    let mut age = input.current_age;
    // Grow balance up to start age (no RMDs yet).
    while age < start_age {
        balance *= 1.0 + r;
        age += 1;
    }
    let mut projection: Vec<YearRmd> = Vec::new();
    let mut total_rmds = 0.0_f64;
    for _ in 0..input.project_years {
        let factor_opt = uniform_lifetime_factor(age);
        let (factor, rmd) = match factor_opt {
            Some(f) => (f, balance / f),
            None => break,
        };
        let start = balance;
        let end = (balance - rmd).max(0.0);
        projection.push(YearRmd {
            age,
            start_balance_usd: start,
            rmd_factor: factor,
            rmd_amount_usd: rmd,
            end_balance_after_rmd_usd: end,
        });
        total_rmds += rmd;
        balance = end * (1.0 + r);
        age += 1;
    }
    RmdReport {
        rmd_start_age: start_age,
        current_factor,
        current_rmd_usd: current_rmd,
        years_until_rmd: years_until,
        projection,
        total_rmds_through_projection_usd: total_rmds,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rmd_start_age_pre_secure_2_0() {
        assert_eq!(rmd_start_age_for_birth_year(1948), 72);
    }

    #[test]
    fn rmd_start_age_secure_2_0_window() {
        assert_eq!(rmd_start_age_for_birth_year(1951), 73);
        assert_eq!(rmd_start_age_for_birth_year(1959), 73);
    }

    #[test]
    fn rmd_start_age_post_2033() {
        assert_eq!(rmd_start_age_for_birth_year(1960), 75);
        assert_eq!(rmd_start_age_for_birth_year(1975), 75);
    }

    #[test]
    fn uniform_factor_known_ages() {
        assert_eq!(uniform_lifetime_factor(73), Some(26.5));
        assert_eq!(uniform_lifetime_factor(80), Some(20.2));
        assert_eq!(uniform_lifetime_factor(90), Some(12.2));
        assert_eq!(uniform_lifetime_factor(100), Some(6.4));
    }

    #[test]
    fn uniform_factor_outside_table() {
        assert!(uniform_lifetime_factor(50).is_none());
        assert!(uniform_lifetime_factor(125).is_none());
    }

    #[test]
    fn compute_pre_rmd_age_no_current_rmd() {
        let r = compute(&RmdInput {
            birth_year: 1955,
            current_age: 65,
            balance_usd: 500_000.0,
            expected_annual_return_pct: 6.0,
            project_years: 20,
        });
        assert_eq!(r.rmd_start_age, 73);
        assert!(r.current_rmd_usd.is_none());
        assert_eq!(r.years_until_rmd, 8);
    }

    #[test]
    fn compute_at_rmd_age_current_rmd() {
        let r = compute(&RmdInput {
            birth_year: 1952,
            current_age: 73,
            balance_usd: 1_000_000.0,
            expected_annual_return_pct: 6.0,
            project_years: 20,
        });
        assert_eq!(r.rmd_start_age, 73);
        // $1M / 26.5 = $37,736
        assert!(r.current_rmd_usd.is_some());
        assert!((r.current_rmd_usd.unwrap() - 37_735.85).abs() < 1.0);
    }

    #[test]
    fn compute_projection_starts_at_rmd_age() {
        let r = compute(&RmdInput {
            birth_year: 1955,
            current_age: 65,
            balance_usd: 500_000.0,
            expected_annual_return_pct: 6.0,
            project_years: 5,
        });
        assert_eq!(r.projection.first().unwrap().age, 73);
    }

    #[test]
    fn compute_projection_year_count() {
        let r = compute(&RmdInput {
            birth_year: 1952,
            current_age: 73,
            balance_usd: 1_000_000.0,
            expected_annual_return_pct: 6.0,
            project_years: 10,
        });
        assert_eq!(r.projection.len(), 10);
    }

    #[test]
    fn compute_projection_balance_after_rmd_decreasing_when_rmd_exceeds_growth() {
        // After ~age 88, RMD divisor < 14 → RMD > 7%, exceeds 6% return.
        let r = compute(&RmdInput {
            birth_year: 1952,
            current_age: 88,
            balance_usd: 500_000.0,
            expected_annual_return_pct: 6.0,
            project_years: 5,
        });
        let first = r.projection[0].end_balance_after_rmd_usd;
        let last_start = r.projection.last().unwrap().start_balance_usd;
        assert!(last_start < first || (last_start - first).abs() < first * 0.5);
    }

    #[test]
    fn compute_total_rmds_sums_projection() {
        let r = compute(&RmdInput {
            birth_year: 1952,
            current_age: 73,
            balance_usd: 1_000_000.0,
            expected_annual_return_pct: 6.0,
            project_years: 5,
        });
        let sum: f64 = r.projection.iter().map(|y| y.rmd_amount_usd).sum();
        assert!((r.total_rmds_through_projection_usd - sum).abs() < 0.01);
    }
}

File: frontend/js/views/rmd_calculator.js
// Required Minimum Distribution calculator (IRS Pub 590-B + SECURE 2.0).
// Reports current-year RMD, years until RMDs start, projection table.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';
import * as enh from '../calc_enhance.js';

export async function renderRmdCalculator(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.rmd_calculator.title">// REQUIRED MINIMUM DISTRIBUTION</span></h1>
        <p class="muted small" data-i18n-html="view.rmd_calculator.intro">
            Per <strong>SECURE 2.0 Act (2022)</strong> and IRS Publication 590-B:
            RMDs begin at <strong>age 73</strong> for those born 1951-1959 and at
            <strong>age 75</strong> for those born 1960+.
            <code>RMD = prior_year_end_balance / Uniform_Lifetime_Factor</code>.
            Missed-RMD penalty: 25% (10% if corrected within 2 years). Embedded Uniform
            Lifetime Table covers ages 72-120 per Pub 590-B Appendix B Table III.
        </p>
        <div class="chart-panel">
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:8px;margin-bottom:12px">
                <label><span class="muted small" data-i18n="view.rmd_calculator.field.birth">Birth year</span>
                    <input type="number" id="rc-birth" step="1" min="1900" max="2100" value="1955" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.rmd_calculator.field.age">Current age</span>
                    <input type="number" id="rc-age" step="1" min="1" max="120" value="73" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.rmd_calculator.field.balance">Retirement balance $</span>
                    <input type="number" id="rc-balance" step="10000" min="0" value="1000000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.rmd_calculator.field.return">Expected return %/yr</span>
                    <input type="number" id="rc-return" step="0.25" min="-20" max="30" value="6" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.rmd_calculator.field.years">Project years</span>
                    <input type="number" id="rc-years" step="1" min="0" max="60" value="20" style="width:100%"></label>
            </div>
            <button class="btn btn-sm primary" id="rc-run" data-shortcut="r" data-i18n="view.rmd_calculator.btn.run">⚡ Compute RMD</button>
            <div id="rc-result"></div>
        </div>
    `;
    mount.querySelector('#rc-run').addEventListener('click', () => runCompute(mount));
    await runCompute(mount);
}

async function runCompute(mount) {
    const result = mount.querySelector('#rc-result');
    const input = {
        birth_year: parseInt(mount.querySelector('#rc-birth').value, 10) || 0,
        current_age: parseInt(mount.querySelector('#rc-age').value, 10) || 0,
        balance_usd: parseFloat(mount.querySelector('#rc-balance').value) || 0,
        expected_annual_return_pct: parseFloat(mount.querySelector('#rc-return').value) || 0,
        project_years: parseInt(mount.querySelector('#rc-years').value, 10) || 0,
    };
    result.innerHTML = `<p class="muted">${esc(t('view.rmd_calculator.status.computing'))}</p>`;
    try {
        const r = await api.request('/rmd-calculator/compute', { method: 'POST', body: JSON.stringify(input) });
        // RMD-per-age bar chart from the projection (the withdrawal ramp).
        const chart = enh.svgBarChart((r.projection || []).map(y => ({ label: String(y.age), value: y.rmd_amount_usd })));
        const untilCls = r.years_until_rmd <= 0 ? '' : 'pos';
        result.innerHTML = `
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:12px;margin-top:1rem">
                <div><div class="muted small">${esc(t('view.rmd_calculator.field.start_age'))}</div>
                    <strong>${r.rmd_start_age}</strong></div>
                <div><div class="muted small">${esc(t('view.rmd_calculator.field.years_until'))}</div>
                    <strong class="${untilCls}">${r.years_until_rmd}</strong></div>
                <div><div class="muted small">${esc(t('view.rmd_calculator.field.current_factor'))}</div>
                    <strong>${r.current_factor == null ? '—' : r.current_factor.toFixed(1)}</strong></div>
                <div><div class="muted small">${esc(t('view.rmd_calculator.field.current_rmd'))}</div>
                    <strong style="font-size:1.3em">${r.current_rmd_usd == null ? '—' : '$' + r.current_rmd_usd.toFixed(0)}</strong></div>
                <div><div class="muted small">${esc(t('view.rmd_calculator.field.total_rmds'))}</div>
                    <strong class="neg">$${(r.total_rmds_through_projection_usd / 1000).toFixed(0)}K</strong></div>
            </div>
            ${chart}
            <div id="rmd-tools" class="ce-toolbar"></div>
            <h2 style="margin-top:1rem">${esc(t('view.rmd_calculator.h2.projection'))}</h2>
            <table class="trades">
                <thead><tr>
                    <th data-i18n="view.rmd_calculator.th.age">Age</th>
                    <th data-i18n="view.rmd_calculator.th.start">Start balance</th>
                    <th data-i18n="view.rmd_calculator.th.factor">Factor</th>
                    <th data-i18n="view.rmd_calculator.th.rmd">RMD</th>
                    <th data-i18n="view.rmd_calculator.th.end">End balance</th>
                </tr></thead>
                <tbody>${(r.projection || []).map(y => `
                    <tr>
                        <td>${y.age}</td>
                        <td>$${(y.start_balance_usd / 1000).toFixed(0)}K</td>
                        <td>${y.rmd_factor.toFixed(1)}</td>
                        <td class="neg">$${y.rmd_amount_usd.toFixed(0)}</td>
                        <td>$${(y.end_balance_after_rmd_usd / 1000).toFixed(0)}K</td>
                    </tr>
                `).join('')}</tbody>
            </table>
        `;
        // Projection export (Copy / CSV). No permalink — id-based inputs.
        enh.mountToolbar(mount.querySelector('#rmd-tools'), {
            viewId: 'rmd-calculator',
            link: false,
            filename: 'rmd-calculator.csv',
            getRows: () => [['age', 'start_balance_usd', 'rmd_factor', 'rmd_amount_usd', 'end_balance_usd'],
                ...(r.projection || []).map(y => [y.age, y.start_balance_usd, y.rmd_factor, y.rmd_amount_usd, y.end_balance_after_rmd_usd])],
        });
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}
